### DISCUSSION TITLE
Help on promise chains and inputs to the model

### DISCUSSION THREAD

**Discussion by luthis1124** [oobabooga/text-generation-webui#2186] [Q&A] [answered]
So 2 questions:

Can anyone syntactically explain what is going on here:

```
            gen_events.append(shared.gradio['Generate'].click(
                ui.gather_interface_values, [shared.gradio[k] for k in shared.input_elements], shared.gradio['interface_state']).then(
                lambda x: (x, ''), shared.gradio['textbox'], [shared.gradio['Chat input'], shared.gradio['textbox']], show_progress=False).then(
                chat.generate_chat_reply_wrapper, shared.input_params, shared.gradio['display'], show_progress=False).then(
                chat.save_history, shared.gradio['mode'], None, show_progress=False)
            )
```

I just need to know how the 'then' blocks are working together, what inputs and outputs are going where, etc. Not what this code actually _does_ at runtime. I can't seem to find documentation for this syntax, I always find myself on JS pages but that is still different syntax to this. Even chatGPT gets confused.

Secondly, and this is the reason for needing the first part, I want to be able to pass arbitrary strings from an extension as a reply to the model. So not necessarily as a button action or whatever. Eg, perhaps I want the model to tell me I received an email. 

The key is here, ie perhaps writing text to the textbox and clicking generate from in the code, or taking the innards from 'click' and running that directly, (I can't append to gen_events from outside server.py) but I haven't been able to find a way yet.

Thanks!

**Comment by OrenRele** [comment, marked as answer]
To answer the first question, .then() is:
`"""
        Triggered after directly preceding event is completed, regardless of success or failure.
        """` 
meaning that something happens on .click(), and then something happens on .then(), regardless if the thing that happened before was successful or not. Note that the callback is like click, it receives a function to call, a list of inputs (that will be passed to the function) and a list of outputs from the function.

Regarding the second question, I think the specific ".then()" that's interesting to you is: `.then(chat.generate_chat_reply_wrapper, ..)` as it is the one that actually sends stuff to the model and does something with the output.

Hope this helps!

**Reply by luthis1124** [reply]
Thanks for replying!

After much testing, I have come to the conclusion that the only way I can achieve this functionality is by using the API.

As far as I can tell, when the button is clicked, those actions in the then() chain are added to an event queue as part of click() and processed at some asynchronous point in time later. I'm not able to add the events manually, so what I will do is use the API (this has been discussed in the Gradio github by others wanting to 'programatically click the button'; keywords for you incase you want to search it yourself).

I have no idea how I'm going to do this yet, but I know it is definitely possible so... I'll see how we go!

Hopefully I'll be able to contribute a nice fun extension soon
